homeassistant/components/tplink/entity.py

```diff
@@ -20,10 +21,16 @@
 def async_refresh_after(
     func: Callable[Concatenate[_T, _P], Awaitable[None]],
 ) -> Callable[Concatenate[_T, _P], Coroutine[Any, Any, None]]:
-    """Define a wrapper to refresh after."""
+    """Define a wrapper to catch exceptions and refresh after calls."""
 
     async def _async_wrap(self: _T, *args: _P.args, **kwargs: _P.kwargs) -> None:
-        await func(self, *args, **kwargs)
+        try:
+            await func(self, *args, **kwargs)
+        except AuthenticationException as ex:
+            raise ConfigEntryAuthFailed from ex
```

Comment:
I don't think this will work here because the service call isn't going to have enough context to start reauth

Reply:
We would need to do `self.coordinator.config_entry.async_start_reauth(self.hass)`

and also `raise HomeAssistantError`